ui/components/compose.jsx

```diff
@@ -115,25 +119,37 @@ module.exports = function(core, config, store) {
 			}
 		},
 
+		onStateChange: function(changes, next) {
+			if (changes.user || (changes.nav && changes.nav.currentText)) {
+				this.onChange(true);
+			}
+
+			next();
+		},
+
 		componentDidMount: function() {
-			this.setPlaceHolder();
+			this.focusInput();
+
+			core.on("statechange", this.onStateChange, 100);
```

Comment:
This is weird. On statechange you are getting re-rendered anyway. You can put the this.onStateChange logic in a componentWillUpdate function.

You also dont need to listen onInput or onPaste... inside componentWillUpdate, you can read the entered text from the DOM before it is overwritten, get the currently selected text from store, run the getMessageText algorithm if needed and call this.setState({userInput}) with result.